clinvar_ingest/cloud/gcs.py

```diff
@@ -1,27 +1,76 @@
 import logging
+import subprocess
+import urllib
 
 from google.cloud import storage
 
 _logger = logging.getLogger(__name__)
 
-client = storage.Client()
 
-
-def parse_blob_uri(uri: str) -> storage.Blob:
+def parse_blob_uri(uri: str, client: storage.Client = None) -> storage.Blob:
     if not uri.startswith("gs://"):
         raise ValueError("Must be a fully qualified URI beginning with gs://")
+    if client is None:
+        client = storage.Client()
     proto, *rest = uri.split("://")
     bucket, *path_segments = rest[0].split("/")
     return storage.Blob(
         name="/".join(path_segments), bucket=storage.Bucket(client=client, name=bucket)
     )
 
 
-def copy_file_to_bucket(local_file_uri: str, remote_blob_uri):
+def copy_file_to_bucket(
+    local_file_uri: str, remote_blob_uri: str, client: storage.Client = None
+):
     """
     Upload the contents of file `local_file_uri` on local filesystem, to `remote_blob_uri` in
     """
     _logger.info(f"Uploading {local_file_uri} to {remote_blob_uri}")
-    blob = parse_blob_uri(remote_blob_uri)
+    if client is None:
+        client = storage.Client()
+    blob = parse_blob_uri(remote_blob_uri, client=client)
     blob.upload_from_filename(client=client, filename=local_file_uri)
     _logger.info(f"Finished uploading {local_file_uri} to {remote_blob_uri}")
+
+
+def blob_writer(
+    blob_uri: str, client: storage.Client = None, binary=True
+) -> storage.Blob:
+    """
+    Returns a file-like object that can be used to write to the blob at `blob_uri`
+    """
+    if client is None:
+        client = storage.Client()
+    blob = parse_blob_uri(blob_uri, client=client)
+    return blob.open("wb" if binary else "w")
+
+
+def http_upload_shell(http_uri: str, blob_uri: str):
+    """
+    Upload the contents of `http_uri` to `blob_uri` using curl and gsutil
+    """
+    p = subprocess.run(
+        ["bash", "-c", f"curl {http_uri} | gcloud storage cp - {blob_uri}"],
+        capture_output=True,
+        check=True,
+    )
+
+    if p.returncode != 0:
```

Comment:
`check=True` forces an exception to be raised if there's a nonzero exit code so I think this is dead code:

```
>>> p = subprocess.run(["curl", "1"], capture_output=True, check=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/bblanken/.pyenv/versions/3.8.10/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['curl', '1']' returned non-zero exit status 7.
>>> p.returncode
0
```

(this is only relevant if we keep this function I guess!)

Reply:
Good catch. 

And I think we should go with the pure python approach instead of subprocessing out to a shell command, and also delete this function.

Reply:
agreed!